File: examples/app-java/src/main/java/pers/xuankai/udptestjava/activity/MQTTActivity.java
package pers.xuankai.udptestjava.activity;

import android.content.Intent;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.dylanc.longan.SystemBarsKt;
import com.gitee.xuankaicat.kmnkt.socket.IMqttSocket;
import com.gitee.xuankaicat.kmnkt.socket.ISocket;
import com.gitee.xuankaicat.kmnkt.socket.OnOpenCallback;
import com.gitee.xuankaicat.kmnkt.socket.utils.CharsetUtils;
import pers.xuankai.udptestjava.BaseActivity;
import pers.xuankai.udptestjava.databinding.ActivityMqttactivityBinding;

import java.nio.charset.Charset;

public class MQTTActivity extends BaseActivity<ActivityMqttactivityBinding> {
    private final IMqttSocket mqtt = IMqttSocket.getMQTT(s -> {
        s.setAddress("10.0.2.2");
        s.setPort(1883);
        CharsetUtils.setInCharset(s, Charset.forName("gb2312"));
        CharsetUtils.setOutCharset(s, Charset.forName("gb2312"));
        s.setUsername("siot");
        s.setPassword("siot");
        s.setInMessageTopic("DeviceTest/000000");
        s.setOutMessageTopic("DeviceTest/123456");
        s.open(new OnOpenCallback(s) {
            @Override
            public void success(@NonNull ISocket socket) {
                super.success(socket);
                socket.startReceive((result, ignore) -> {
                    getBinding().textView.setText(result);
                    return true;
                });
            }
        });
        return null;
    });

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        SystemBarsKt.immerseStatusBar(this, true, true);

        ActivityMqttactivityBinding binding = getBinding();

        binding.btn.setOnClickListener(v -> {
            String sendText = binding.editText.getText().toString();
            if(sendText.equals("")) return;

            mqtt.send(sendText);
            binding.textView.setText("等待数据...");
            mqtt.startReceive((result, ignore) -> {
                binding.textView.setText(result);
                return false;
            });
        });

        /*
          以一个收到status决定亮灭的小灯为例，按下发送开启时小灯亮，按下发送关闭时小灯灭
         */

        binding.btnOpen.setOnClickListener(v -> {
            String sendText = "{cmd:\"set\", status:1}";
            mqtt.send(sendText);
        });

        binding.btnClose.setOnClickListener(v -> {
            String sendText = "{cmd:\"set\", status:0}";
            mqtt.send(sendText);
        });

        binding.btnUdp.setOnClickListener(v -> {
            Intent intent = new Intent(MQTTActivity.this, UDPActivity.class);
            startActivity(intent);
        });
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mqtt.stopReceive();
        mqtt.close();
    }
}

File: examples/app/src/main/java/pers/xuankai/udptest/activity/MQTTActivity.kt
package pers.xuankai.udptest.activity

import android.os.Bundle
import com.dylanc.longan.immerseStatusBar
import com.dylanc.longan.startActivity
import com.gitee.xuankaicat.kmnkt.socket.dsl.mqtt
import com.gitee.xuankaicat.kmnkt.socket.open
import pers.xuankai.udptest.BaseActivity
import pers.xuankai.udptest.databinding.ActivityMqttactivityBinding
import java.nio.charset.Charset

class MQTTActivity : BaseActivity<ActivityMqttactivityBinding>() {
    private val mqtt = mqtt {
        address = "10.0.2.2"
        port = 1883
        inCharset = Charset.forName("gb2312")
        outCharset = Charset.forName("gb2312")
        username = "xuankai"
        password = "xuankai"
        inMessageTopic = "DeviceTest/111111"
        outMessageTopic = "DeviceTest/123456"
        open {
            success {
                it.startReceive { str, _ ->
                    binding.textView.text = str
                    return@startReceive true
                }
            }
            failure {
                Log.i("zyq","链接失败，1秒后重连")
                Thread.sleep(1000)
                //communicate.open(this)
                //开启连接失败时执行
                return@failure true//是否继续尝试连接
            }
            loss {
                Log.v("MQTT", "失去连接，尝试重连")
                return@loss true
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        immerseStatusBar()

        binding.btn.setOnClickListener {
            val sendText = binding.editText.text.toString()
            if(sendText.isEmpty()) return@setOnClickListener

            mqtt.send(sendText)
            binding.textView.text = "等待数据..."
            mqtt.startReceive { str, _ ->
                binding.textView.text = str
                return@startReceive false
            }
        }

        /**
         * 以一个收到status决定亮灭的小灯为例，按下发送开启时小灯亮，按下发送关闭时小灯灭
         */

        binding.btnOpen.setOnClickListener {
            val sendText = "{cmd:\"set\", status:1}"
            mqtt.send(sendText)
        }

        binding.btnClose.setOnClickListener {
            val sendText = "{cmd:\"set\", status:0}"
            mqtt.send(sendText)
        }

        binding.btnUdp.setOnClickListener {
            startActivity<UDPActivity>()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        mqtt.stopReceive()
        mqtt.close()
    }
}